src/scripts/main.js

```diff
@@ -1,3 +1,253 @@
+
+  return false;
+}
+
+function generateRandomCell() {
+  if (!hasEmptyCell()) {
+    return;
+  }
+
+  let found = false;
+
+  while (!found) {
+    const randomCellNum = Math.random() <= 0.1 ? 4 : 2;
+    const r = Math.floor(Math.random() * rows);
+    const c = Math.floor(Math.random() * columns);
+
+    if (board[r][c] === 0) {
+      board[r][c] = randomCellNum;
+      found = true;
+      updateCell(document.getElementById(`${r}-${c}`), randomCellNum);
+    }
+  }
+}
+
+function checkWin() {
+  if (winCell.length > 0) {
+    messageWin.classList.remove('hidden');
+  }
+}
+
+function checkLose() {
+  if (hasEmptyCell()) {
+    return false;
+  }
+
+  for (let r = 0; r < rows; r++) {
+    for (let c = 0; c < columns; c++) {
+      if (board[r][c] === board[r][c + 1]) {
+        return false;
+      }
+    }
+  }
+
+  for (let c = 0; c < columns; c++) {
+    const row = [board[0][c], board[1][c], board[2][c], board[3][c]];
+
+    for (let r = 0; r < row.length; r++) {
+      if (row[r] === row[r + 1]) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+function showLoseMessage() {
+  messageLose.classList.remove('hidden');
+}
+
+function updateCell(cell, num) {
+  cell.innerText = '';
+  cell.classList.value = '';
+  cell.classList.add('field-cell');
+
+  if (num > 0) {
+    cell.innerText = num;
+
+    cell.classList.add(`field-cell--${num}`);
+  }
+}
+
+document.addEventListener('keydown', (e) => {
+  const arrowLeft = 'ArrowLeft';
+  const arrowRight = 'ArrowRight';
+  const arrowUp = 'ArrowUp';
+  const arrowDown = 'ArrowDown';
+
+  switch (e.code) {
+    case arrowLeft:
+      slideLeft();
+      break;
+    case arrowRight:
+      slideRight();
+      break;
+    case arrowUp:
+      slideUp();
+      break;
+    case arrowDown:
+      slideDown();
+      break;
+  }
+
+  generateRandomCell();
+  scoreCell.innerText = score;
+  checkWin();
+
+  if (checkLose()) {
+    showLoseMessage();
+  }
+});
+
+function filterZero(row) {
+  return row.filter(num => num !== 0);
+}
+
+function slide(row) {
+  let currentRow = filterZero(row);
+
+  for (let i = 0; i < currentRow.length - 1; i++) {
+    if (currentRow[i] === currentRow[i + 1]) {
+      currentRow[i] *= 2;
+      currentRow[i + 1] = 0;
+      score += currentRow[i];
+    }
+  }
+
+  currentRow = filterZero(currentRow);
+
+  while (currentRow.length < columns) {
+    currentRow.push(0);
+  }
+
+  return currentRow;
+}
+
+function slideLeft() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row = slide(row);
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      updateCell(document.getElementById(`${r}-${c}`), board[r][c]);
+    }
+  }
+}
+
+function slideRight() {
+  for (let r = 0; r < rows; r++) {
+    let row = board[r];
+
+    row.reverse();
+
+    row = slide(row);
+
+    row.reverse();
+    board[r] = row;
+
+    for (let c = 0; c < columns; c++) {
+      updateCell(document.getElementById(`${r}-${c}`), board[r][c]);
+    }
```

Comment:
you are repeating this in `slideLeft`, you can move it to separate reusable func